ocaml/idl/datamodel.ml

```diff
@@ -8414,3 +8414,44 @@ let extra_permissions =
     (Task.extra_permission_task_destroy_any, _R_POOL_OP)
     (* only POOL_OP can destroy any tasks *)
   ]
+
+module StringMap = Map.Make (String)
+
+type obj_states = {
+    obj_state: Lifecycle.state
+  ; msg_states: Lifecycle.state StringMap.t
+  ; fld_states: Lifecycle.state StringMap.t
+}
+
+let all_lifecycles =
+  let parse_with ~name lifecycle =
+    try Lifecycle.from lifecycle
```

Comment:
Should this use a `result` type instead, or do you want to preserve the location that raised the invalid message in stacktraces (if it is useful then it should be preserved and using exceptions here is better)

Reply:
This code is replaced when the lifecycle state is integrated into the datamodel, I don't think it's important enough to change it for the few commits it's used